### PR TITLE
[JENKINS-40098] Bundle naming strategy should be able to specify an instance type

### PR DESCRIPTION
**Pull request by batmat**
[JENKINS-40098](https://issues.jenkins-ci.org/browse/JENKINS-40098)

Refactored the existing code to use the same logic for choosing the bundle filename, and introduced a change enabling both overriding through using a sysprop, or implementing a new Extension Point.

The approach with Extension Points should enable more dynamic needs, when the sysprop one will be immediately usable by anyone without having to develop an extension.

Note: if found, the extension is preferred over the sysprop.

@reviewbybees esp. @jglick who pointed me to SupportProvider.getName() I had missed in the first attempt. 

Supersedes #91

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
+package com.cloudbees.jenkins.support;
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: (c) header please

File: src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        final ExtensionList<BundleNameInstanceTypeProvider> all = ExtensionList.lookup(BundleNameInstanceTypeProvider.class);
+        final int extensionCount = all.size();
+        if (extensionCount < 1) {
+            LOGGER.fine("No alternative strategy provided for support bundle prefixing.");
+            return DEFAULT_STRATEGY;
+        }
+
+        if (all.size() > 1) {
+            if (LOGGER.isLoggable(Level.WARNING)) {
+                LOGGER.warning("{0} implementations found for support bundle prefix naming strategy. " +
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: :bug: you are missing a parameter here.
NIT: from a stylistic point I think its better to always use Logger.log(Level.XXX ...) rather than the Logger.xxx(...)

File: src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    @Nonnull
+    static BundleNameInstanceTypeProvider getInstance() {
+        final ExtensionList<BundleNameInstanceTypeProvider> all = ExtensionList.lookup(BundleNameInstanceTypeProvider.class);
+        final int extensionCount = all.size();
+        if (extensionCount < 1) {
+            LOGGER.fine("No alternative strategy provided for support bundle prefixing.");
+            return DEFAULT_STRATEGY;
+        }
+
+        if (all.size() > 1) {
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: why not use `extensionCount`

File: src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+     * Aims to provide informational data about the generated bundles.
+     *
+     * <p>
+     * <p><b>Will be used for file name generation, so avoid funky characters.
+     * Please stay in <code>[a-zA-Z-_]</code></b>. Also consider the file name length, you probably want to be defensive
+     * and not return crazily long strings. Something below 20 characters or so might sound reasonable.</p>
+     *
+     * @return the instance type specification to be used for generated support bundles.
+     */
+    @Nonnull
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: is this really the annotations that this code already uses.... If not then swtich.

File: src/test/java/com/cloudbees/jenkins/support/BundleNamePrefixTest.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    public JenkinsRule rule = new JenkinsRule();
+
+    @Test
+    public void checkOriginalBehaviour() throws Exception {
+        assertThat(SupportPlugin.getBundleFileName(), startsWith("support_" + CURRENT_YEAR));
+    }
+
+    @Test
+    public void checkWithOneProvider() throws Exception {
+        assertThat(SupportPlugin.getBundleFileName(), startsWith("support_pouet_" + CURRENT_YEAR));
```
Comment (amuniz) [inline src/test/java/com/cloudbees/jenkins/support/BundleNamePrefixTest.java]: Oh, ok, I see the method name match with the `@TestExtension` parameter.

File: src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+ * @see SupportProvider#getName() for prefixing.
+ */
+public abstract class BundleNameInstanceTypeProvider implements ExtensionPoint {
+
+    @VisibleForTesting
+    static final String SUPPORT_BUNDLE_NAMING_INSTANCE_SPEC_PROPERTY = SupportPlugin.class.getName() + ".instanceType";
+
+    private static final Logger LOGGER = Logger.getLogger(BundleNameInstanceTypeProvider.class.getName());
+
+    private static final BundleNameInstanceTypeProvider DEFAULT_STRATEGY = new BundleNameInstanceTypeProvider() {
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: why not have this as an extension with very low priority and avoid the extra complexity below?

File: src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java [hunk 7]
```diff
@@ -763,6 +774,33 @@ public boolean accept(File dir, String name) {
             }
         }
 
+        private void cleanupOldBundles(File bundleDir, File justGenerated) {
+            thread.setName(String.format("%s periodic bundle generator: tidying old bundles since %s",
+                    SupportPlugin.class.getSimpleName(), new Date()));
+            File[] files = bundleDir.listFiles(new FilenameFilter() {
+                public boolean accept(File dir, String name) {
+                    return name.startsWith(SupportPlugin.getBundlePrefix()) && name.endsWith(".zip");
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java]: this will fail to clean up old bundles if the provider is changed.

Why not just check for "support*.zip"?

File: src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            LOGGER.fine("No alternative strategy provided for support bundle prefixing.");
+            return DEFAULT_STRATEGY;
+        }
+
+        if (all.size() > 1) {
+            if (LOGGER.isLoggable(Level.WARNING)) {
+                LOGGER.warning("{0} implementations found for support bundle prefix naming strategy. " +
+                        "Can be only 0 or 1. Choosing the first found.");
+                for (BundleNameInstanceTypeProvider bundlePrefixProvider : all) {
+                    LOGGER.log(Level.WARNING, "class '{0}' found", bundlePrefixProvider.getClass().getName());
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: :bug:  `'` needs to be escaped other wise this will log `"class {0} found`"
https://docs.oracle.com/javase/7/docs/api/index.html?java/text/MessageFormat.html

File: src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java (hunk 6 above)
Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: Just because I wasn't absolutely sure about the behaviour of the priority, going to check, and if that works and the doc confirms, will do.

File: src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java (hunk 7 above)
Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java]: Because the SupportProvider.getName() could make that pattern totally different anyway, files could be named `whatever-thedate.zip`.

A different approach might be to just select every zip files like `$JENKINS_HOME/support/*.zip`, but this is a bit risky if someone has taken the habit to put unrelated files there...

**Comment by batmat** [comment]
@jtnord I believe I addressed all your comments, but https://github.com/jenkinsci/support-core-plugin/pull/92#discussion_r90266258. For that one, I guess the only "risk" is to have uncleaned lefovers that users may have to clean up manually. The alternative is to work more largely, but we take then IMO an even more risky approach: wiping out unrelated zip files. WDYT?

If you have a better idea or think this is fine to just consider nobody should have put things in that directory (which, well, I would agree on...) then we can make it simple(r).

**Comment by amuniz** [comment]
:bee:

File: src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+/**
+ * <p>Extension point allowing to customize the support bundle naming strategy.</p>
+ * <p>
+ * It will work the following way:
+ * </p>
+ * <ol>
+ * <li>If an implementation of {@link BundleNameInstanceTypeProvider} is found, it will be used.<br>
+ * <strong>WARNING: </strong>if many are found, then a warning will be issued, and the first extension found will
+ * be used.</li>
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: So, where is this nondefault implementation? For any newly introduced API I expect to see a PR from another plugin repository using a `SNAPSHOT` dependency which consumes that API. (Obviously public use cases are preferred whenever possible, but note that a PR in a private repository may refer in its summary etc. to a public PR, in which case GitHub automatically creates a backlink in the public PR visible to users who have permission to view the private repo.)

File: src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+     * Please stay in <code>[a-zA-Z-_]</code></b>. Also consider the file name length, you probably want to be defensive
+     * and not return crazily long strings. Something below 20 characters or so might sound reasonable.</p>
+     *
+     * @return the instance type specification to be used for generated support bundles.
+     */
+    @NonNull
+    public abstract String getInstanceType();
+
+    // We want this to be always picked up last in case others are found Double.MIN_VALUE does NOT work
+    @Extension(ordinal = -1000000)
```
Comment (christ66) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: `@Extension(ordinal = Integer.MAX_VALUE)`

Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: Well, why not, but this is still weird, since the type is `double` anyway...

File: src/main/java/com/cloudbees/jenkins/support/SupportCommand.java [hunk 11]
```diff
@@ -107,7 +98,7 @@ protected int run() throws Exception {
         }
 
         public OutputStream call() throws IOException {
-            File f = File.createTempFile(filename, ".zip");
+            File f = File.createTempFile(filename, "");
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/SupportCommand.java]: This is wrong. Even if `filename` ends in `.zip`, you will end up with something like this:

```
$ jrunscript -e 'println(java.io.File.createTempFile("foo.zip", ""))'
/tmp/foo.zip6023275810588304403
```

File: src/main/java/com/cloudbees/jenkins/support/SupportCommand.java [hunk 12]
```diff
@@ -107,7 +98,7 @@ protected int run() throws Exception {
...
             System.err.println("Creating: " + f);
             return new RemoteOutputStream(new FileOutputStream(f));
         }
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/SupportCommand.java]: BTW not introduced in this PR, but while you are here please delete the bogus `checkRoles` implementation and extend `MasterToSlaveCallable`.

File: src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java [hunk 13]
```diff
@@ -563,6 +561,46 @@ public void run() {
...
+    @NonNull
+    public static String getBundleFileName() {
+        StringBuilder filename = new StringBuilder();
+        filename.append(getBundlePrefix());
+
+        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd_HH.mm.ss");
+        dateFormat.setTimeZone(TimeZone.getTimeZone("GMT"));
+        filename.append("_").append(dateFormat.format(new Date()));
+
+        filename.append(".zip");
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java]: As noted above, better to have the extension appended by the caller.

File: src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+        final BundleNameInstanceTypeProvider chosen = all.get(0);
+        if (extensionCount > 1) {
+            LOGGER.log(Level.INFO, "Using {0} as BundleNameInstanceTypeProvider implementation", chosen.getClass().getName());
+        }
+        return chosen;
+    }
+
+    /**
+     * Returns the <strong>non-null</strong> instance type to be used for generated support bundle names.
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: You need to document that the empty string is allowed, and in fact the default.

File: src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java [hunk 15]
```diff
@@ -563,6 +561,46 @@ public void run() {
...
+     * @return the prefix of the bundle name.
+     */
+    private static String getBundlePrefix() {
+        String filename = "support"; // default bundle filename
+        final SupportPlugin instance = getInstance();
+        if (instance != null) {
+            SupportProvider supportProvider = instance.getSupportProvider();
+            if (supportProvider != null) {
+                // let the provider name it
+                filename = supportProvider.getName();
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java]: Still leaves me wondering about why we have both systems. In the case of a vendor Acme implementing both extension points in a Blaster product, we will get something like

    acme-support_acme-blaster_2016-12-07_16.12.47.zip

which seems redundant.

File: src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java (hunk 7 above)
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java]: > if someone has taken the habit to put unrelated files there

Well don’t do that then.

I agree it would be better to just delete all `*.zip`.

File: src/main/java/com/cloudbees/jenkins/support/SupportCommand.java (hunk 11 above)
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/SupportCommand.java]: Though if we are always adding the current date down to second resolution anyway, then perhaps we can omit the generated suffix and just save `${java.io.tmpdir}/${filename}`.

File: src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java (hunk 15 above)
Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java]: Well, in that case, that vendor would just want to not put Acme everywhere. Which would be more:

`acme-support_blaster_2016-12-07_16.12.47.zip`

which seems less redundant than your example?

File: src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+                        "Choosing the first found among the following:", extensionCount);
+                for (BundleNameInstanceTypeProvider nameProvider : all) {
+                    LOGGER.log(Level.WARNING, "Class {0} found", nameProvider.getClass().getName());
+                }
+            }
+        }
+
+        final BundleNameInstanceTypeProvider chosen = all.get(0);
+        if (extensionCount > 1) {
+            LOGGER.log(Level.INFO, "Using {0} as BundleNameInstanceTypeProvider implementation", chosen.getClass().getName());
```
Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: fixed

File: src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java (hunk 14 above)
Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: fixed. I actually explained it's expected to be not empty (contrary to the default value)

File: src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+        }
+        return chosen;
+    }
+
+    /**
+     * Returns the <strong>non-null</strong> instance type to be used for generated support bundle names.
+     * Aims to provide informational data about the generated bundles.
+     * <p>
+     * <p>
+     * <p><b>Will be used for file name generation, so avoid funky characters.
```
Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: fixed

File: src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+     * Please stay in <code>[a-zA-Z-_]</code></b>. Also consider the file name length, you probably want to be defensive
```
Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: fixed

File: src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java (hunk 10 above)
Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: fixed, though I don't really like it. But in the end with the comment above, ppl puzzled by it should understand.

File: src/main/java/com/cloudbees/jenkins/support/SupportCommand.java (hunk 12 above)
Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/SupportCommand.java]: done

File: src/main/java/com/cloudbees/jenkins/support/SupportCommand.java (hunk 11 above)
Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/SupportCommand.java]: Done. I settled on using 

```java
Path path = Files.createFile(Paths.get(System.getProperty("java.io.tmpDir"), filename));
```

File: src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java (hunk 13 above)
Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java]: so, finally used `filename` directly instead of a temp file created by the dedicated API.

File: src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java (hunk 7 above)
Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java]: done

File: src/main/java/com/cloudbees/jenkins/support/SupportCommand.java [hunk 19]
```diff
@@ -99,23 +91,18 @@ protected int run() throws Exception {
...
-            return new RemoteOutputStream(new FileOutputStream(f));
-        }
-
-        /** {@inheritDoc} */
         @Override
-        public void checkRoles(RoleChecker checker) throws SecurityException {
-            // TODO: do we have to verify some role?
+        public OutputStream call() throws IOException {
+            Path path = Files.createFile(Paths.get(System.getProperty("java.io.tmpDir"), filename));
+            System.err.println("Creating: " + path);
```
Comment (christ66) [inline src/main/java/com/cloudbees/jenkins/support/SupportCommand.java]: 🐛 `System.err`

Comment (christ66) [inline src/main/java/com/cloudbees/jenkins/support/SupportCommand.java]: Ah wait this is for the SupportCommand, nvm

File: src/main/java/com/cloudbees/jenkins/support/SupportCommand.java [hunk 20]
```diff
@@ -90,23 +91,18 @@ protected int run() throws Exception {
...
-            System.err.println("Creating: " + f);
-            return new RemoteOutputStream(new FileOutputStream(f));
-        }
-
-        /** {@inheritDoc} */
         @Override
-        public void checkRoles(RoleChecker checker) throws SecurityException {
-            // TODO: do we have to verify some role?
+        public OutputStream call() throws IOException {
+            Path path = Files.createFile(Paths.get(System.getProperty("java.io.tmpDir"), filename));
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/SupportCommand.java]: It is `tmpdir` not `tmpDir`. Did you try running this command?

File: src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+
+    private static final Logger LOGGER = Logger.getLogger(BundleNameInstanceTypeProvider.class.getName());
+
+    @NonNull
+    static BundleNameInstanceTypeProvider getInstance() {
+        final ExtensionList<BundleNameInstanceTypeProvider> all = ExtensionList.lookup(BundleNameInstanceTypeProvider.class);
+        final int extensionCount = all.size();
+
+        if (extensionCount > 2) {
+            if (LOGGER.isLoggable(Level.WARNING)) {
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/BundleNameInstanceTypeProvider.java]: We log `INFO` by default. So no, it is not terribly useful, but harmless I suppose.

**Comment by batmat** [comment]
Fixed @jglick yes I tried it locally using jrunscript, but didn't adjusted after fixing the jrunscript. Sorry for the noise.

**Comment by jglick** [comment]
> I tried it locally using jrunscript

Not sure what you mean. You need to

    mvn hpi:run &
    wget -O /tmp/cli.jar http://localhost:8080/jenkins/jnlpJars/jenkins-cli.jar
    java -jar /tmp/cli.jar -s http://localhost:8080/jenkins/ support

File: src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java [hunk 22]
```diff
@@ -763,6 +770,38 @@ public boolean accept(File dir, String name) {
...
+                    SupportPlugin.class.getSimpleName(), new Date()));
+            File[] files = bundleDir.listFiles(new FilenameFilter() {
+                public boolean accept(File dir, String name) {
+                    return name.endsWith(".zip");
+                }
+            });
+            long pivot = System.currentTimeMillis();
+            for (long l = 1; l * 2 > 0; l *= 2) {
+                boolean seen = false;
+                for (File f : files) {
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java]: NPE if `listFiles` returns null, as it is permitted to do even if you believe it currently exists.

Comment (batmat) [inline src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java]: @jglick I just moved that code in a dedicated method to clarify the code, didn't touch it. But will happily file a separate PR about it.
Thanks for the review !

**Comment by christ66** [comment]
@batmat @jglick @jtnord I am wondering if this PR should be reverted. This only makes the naming of the support bundles significantly longer.

Before this PR my support bundles use to be:

```
cloudbees-support_2017-02-02_18.16.45
```

which is already long enough and unique (as in the time it was created IMO is random and unique enough), but now they are:

```
cloudbees-support_UNKNOWN-localhost_8080_jenkins_2017-01-24_20.02.19
```

Having this long of a name for a folder I do not believe is right. Just a lot of extra noise which does not need to exist in the name.

**Comment by jtnord** [comment]
not sure why that comes up as `UNKNOWN` but the name is really really really useful when you have lots of bundles from the entire cluster in a support ticket that are updated at various occasions during the lifecycle...
The fact it says` cloudbees-support` would possibly imply you are running a cloudbees distribution without some required plugins that should be installed by default on every instance?
